File: Sources/PDFDocument-Extension.m
#import "PDFDocument-Extension.h"

@implementation PDFDocument (Extension)
+ (instancetype)documentWithFilePath:(NSString*)path
{
    return [[PDFDocument alloc] initWithURL:[NSURL fileURLWithPath:path]];
}

+ (instancetype)documentWithMergingPDFFiles:(NSArray<NSString*>*)paths
{
    if (paths.count == 0) {
        return nil;
    }
    
    PDFDocument *doc = [PDFDocument documentWithFilePath:paths[0]];
    if (!doc) {
        return nil;
    }
    
    NSUInteger pageCount = doc.pageCount;
    for (NSUInteger i=1; i<paths.count; i++) {
        PDFDocument *insertedDoc = [PDFDocument documentWithFilePath:paths[i]];
        if (!insertedDoc) {
            return nil;
        }
        for (NSUInteger j=0; j<insertedDoc.pageCount; j++) {
            [doc insertPage:[insertedDoc pageAtIndex:j] atIndex:pageCount];
            pageCount++;
        }
    }
    
    return doc;
}

- (void)appendPage:(PDFPage*)page
{
    [self insertPage:page atIndex:self.pageCount];
}

- (PDFPageBox*)pageBoxAtIndex:(NSUInteger)index
{
    return [self pageAtIndex:index].pageBox;
}

+ (void)fillBackgroundOfPdfFilePath:(NSString*)path withColor:(NSColor*)fillColor
{
    PDFDocument *doc = [PDFDocument documentWithFilePath:path];
    NSUInteger pageCount = doc.pageCount;

    CGColorRef fillColorRef;
    
    if ([fillColor respondsToSelector:@selector(CGColor)]) {
        fillColorRef = fillColor.CGColor;
    } else {
        CGFloat components[fillColor.numberOfComponents];
        [fillColor getComponents:(CGFloat*)&components];
        fillColorRef = (CGColorRef)CGColorCreate(fillColor.colorSpace.CGColorSpace, components);
    }

    for (NSUInteger i = 0; i < pageCount; i++) {
        CGPDFPageRef pdfPageRef = [doc pageAtIndex:i].pageRef;
        const CGRect mediaBoxRect = CGPDFPageGetBoxRect(pdfPageRef, kCGPDFMediaBox);

        CGContextRef contextRef = CGPDFContextCreateWithURL((CFURLRef)[NSURL fileURLWithPath:path], &mediaBoxRect, NULL);

        CGPDFContextBeginPage(contextRef, NULL);

        CGContextSaveGState(contextRef);
        
        CGContextSetFillColorWithColor(contextRef, fillColorRef);
        CGRect drawRect = CGRectMake(mediaBoxRect.origin.x-1, mediaBoxRect.origin.y-1, mediaBoxRect.size.width+2, mediaBoxRect.size.height+2);
        CGContextFillRect(contextRef, drawRect);
        CGContextDrawPDFPage(contextRef, pdfPageRef);
        
        CGContextRestoreGState(contextRef);

        CGPDFContextEndPage(contextRef);
        CGContextRelease(contextRef);
    }
}


@end

File: Sources/PDFDocument-Extension.swift
import Quartz

extension PDFDocument {
    @objc convenience init?(filePath path: String) {
        self.init(url: URL(fileURLWithPath: path))
    }

    @objc convenience init?(merging paths: [String]) {
        let docs = paths.map { PDFDocument(filePath: $0) }
        if docs.contains(nil) {
            return nil
        }

        self.init()
        self.append(docs.compactMap{$0})
    }

    convenience init(merging documents: [PDFDocument]) {
        self.init()
        self.append(documents)
    }

    var pages: [PDFPage] {
        return (0..<self.pageCount).compactMap{ self.page(at: $0) }
    }
    
    func append(_ page: PDFPage) {
        self.insert(page, at: self.pageCount)
    }

    func append(_ document: PDFDocument) {
        document.pages.forEach { self.append($0) }
    }

    func append(_ documents: [PDFDocument]) {
        documents.forEach { self.append($0) }
    }
    
    // 1ページのみの PDF の MediaBox の背景を指定された色で塗りつぶす（複数ページPDFは未対応）。
    // 日本語の埋め込みテキストは壊れてしまう。
    @objc class func fillBackground(of path: String, with fillColor: NSColor) {
        guard let doc = PDFDocument(filePath: path) else { return }
        let fillColorRef = fillColor.cgColor

        for page in doc.pages {
            guard let pdfPageRef = page.pageRef else { return }
            var mediaBoxRect = pdfPageRef.getBoxRect(.mediaBox)

            guard let contextRef = CGContext(URL(fileURLWithPath: path) as CFURL, mediaBox: &mediaBoxRect, nil) else { return }

            contextRef.beginPDFPage(nil)
            contextRef.saveGState()
            contextRef.setFillColor(fillColorRef)

            let drawRect = CGRect(x: mediaBoxRect.origin.x - 1,
                                  y: mediaBoxRect.origin.y - 1,
                                  width: mediaBoxRect.size.width + 2,
                                  height: mediaBoxRect.size.height + 2)

            contextRef.fill(drawRect)
            contextRef.drawPDFPage(pdfPageRef)
            contextRef.restoreGState()
            contextRef.endPDFPage()
        }
    }

}
